src/include/server/mir/scene/text_input_hub.h

```diff
@@ -97,6 +98,48 @@ struct TextInputState
     std::optional<TextInputContentPurpose> content_purpose;
 };
 
+class CopyableWlArray
+{
+public:
+    explicit CopyableWlArray(wl_array* in_data)
+        : data_{copy(in_data)}
+    {
+    }
+
+    CopyableWlArray(CopyableWlArray const& other)
+        : data_{copy(other.data_)}
+    {
+    }
+
+    CopyableWlArray& operator=(CopyableWlArray const& other)
+    {
+        data_ = copy(other.data_);
+        return *this;
+    }
+
+    ~CopyableWlArray()
+    {
+        if (data_)
+            wl_array_release(data_);
```

Comment:
AFAICS, `data_ != nullptr` is an invariant. (There is no default constructor, and the converting constructor will crash if called with `nullptr)`